File: code/ch11/11.4.2_checkinclusion.cpp
class Solution {
    public:
        bool checkInclusion(string s1, string s2) {
            vector<int> cnts(256, 0);
            for (char ch: s1)
                cnts[ch]++;

            int left = s1.size();
            for (int i = 0, j = 0; j < s2.size(); j++) {
                if (cnts[s2[j]]-- > 0)
                    left--;

                while (left == 0) {
                    if (j + 1 - i == s1.size())
                        return true;
                    if (++cnts[s2[i++]] > 0)
                        left++;
                }
            }

            return false;
        }
};

File: code/ch11/11.4.2_checkinclusion.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if (len(s1) > len(s2)):
            return False
        
        list1 = [0 for i in range(26)]
        list2 = [0 for i in range(26)]
        
        for i in range(len(s1)):
            list1[ord(s1[i]) - ord('a')] += 1
            list2[ord(s2[i]) - ord('a')] += 1
            
        count = 0
        for i in range(26):
            if list1[i] == list2[i]:
                count += 1
                
        for i in range(len(s2) - len(s1)):
            right = ord(s2[i + len(s1)]) - ord('a')
            left = ord(s2[i]) - ord('a')
            if count == 26:
                return True
            
            list2[right] += 1
            if list2[right] == list1[right]:
                count += 1
            elif list2[right] == list1[right] + 1:
                count -= 1
                
            list2[left] -= 1
            if list2[left] == list1[left]:
                count += 1
            elif list2[left] == list1[left] - 1:
                count -= 1
                
        return count == 26
